### PR TITLE
feat(texture): support selecting textures

### PR DESCRIPTION
**Pull request by mspivak-actionengine**

### CODE REVIEW

File: src/components/debug-panel/debug-panel.tsx [hunk 1]
```diff
@@ -159,6 +204,62 @@ export const DebugPanel = ({ onClose }: DebugPanelProps) => {
...
+              imageWidth={TEXTURE_ICON_SIZE}
+              imageHeight={TEXTURE_ICON_SIZE}
+            />
+            <PickPanePanel
+              pickPaneSetName={PickPaneSetName.uvDebugTexture}
+              group={TEXTURE_GROUP_CUSTOM}
+              panelLength={5}
+              imageWidth={TEXTURE_ICON_SIZE}
+              imageHeight={TEXTURE_ICON_SIZE}
+            />
```
Comment (belom88) [inline src/components/debug-panel/debug-panel.tsx:222]: Why they devided in 2 groups? It is one list of images

File: src/components/pick-pane-panel/pick-pane-panel.tsx [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    start: number,
+    options: PickPanePanelOptions
+  ): JSX.Element => {
+    const elements: JSX.Element[] = [];
+
+    const startIndex = start >= array.length ? array.length - 1 : start;
+    const endIndex =
+      startIndex + options.panelLength >= array.length
+        ? array.length
+        : startIndex + options.panelLength;
```
Comment (belom88) [inline src/components/pick-pane-panel/pick-pane-panel.tsx:143]: Very weird logic. What do you want to do here?

File: src/components/pick-pane-panel/pick-pane-panel.tsx [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+
+    return <TexturePanel>{elements}</TexturePanel>;
+  };
+
+  return (
+    <>
+      {renderTextureArray(imageArray, {
+        panelLength: panelLength,
+        selectedKey: imagePickedKey,
+      })}
```
Comment (belom88) [inline src/components/pick-pane-panel/pick-pane-panel.tsx:184]: Write JSX here. No sense to call just one function.

**Comment by dsavinov-actionengine** [comment]
@belom88 please review

File: src/redux/slices/pick-pane-slice.ts [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+    pickPaneSetName: string;
+    pickPane: IPickPane;
+    setCurrent: boolean;
+    index?: number;
+  }
+>("addPickPane", async ({ pickPaneSetName, pickPane, setCurrent, index }) => {
+  // if setCurrent === true, we need to fetch the content
+  // if setCurrent === false, it's enough to fetch the icon only.
+  // The content will be fetched when the user pick a pane.
+  await pickPane.fetchPickPane(setCurrent);
```
Comment (belom88) [inline src/redux/slices/pick-pane-slice.ts:144]: Fetch logic should not be in this slice. The component should initiate loading by url when the selected texture has been changed. As the slice is generic, fetch is not needed for basemaps etc.

**Review by belom88 — CHANGES_REQUESTED** [review]

File: src/components/upload-panel/upload-panel.tsx [hunk 5]
```diff
@@ -110,31 +109,31 @@ export const UploadPanel = ({
...
-    <UploadPanelItem
-      title={"Upload Bookmarks"}
-      onCancel={onCancel}
-    >
+    <UploadPanelItem title={title} onCancel={onCancel}>
       <UploadInput
         ref={inputRef}
         id={UPLOAD_INPUT_ID}
         type="file"
+        multiple={multipleFiles ? true : undefined}
```
[the remark this answers is not in the document]
Comment (mspivak-actionengine) [reply src/components/upload-panel/upload-panel.tsx:136]: Done.

File: src/redux/slices/pick-pane-slice.ts [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+ *  PickPaneSet "uvDebugTexture"
+ *    [
+ *      {group:"HQ", icon:"6", id:"k60", custom:false, image: {}},
+ *      {group:"HQ", icon:"7", id:"k61", custom:true,  image: {}},
+ *      {group:"LQ", icon:"8", id:"k70", custom:false, image: {}},
+ *      {group:"LQ", icon:"9", id:"k71", custom:true,  image: {}},
+ *    ]
+ *    pickedId: "k70"
+ */
+interface IPickPaneSet {
```
[the remark this answers is not in the document]
Comment (mspivak-actionengine) [reply src/redux/slices/pick-pane-slice.ts]: Done.

File: src/components/deck-gl-wrapper/deck-gl-wrapper.tsx [hunk 7]
```diff
@@ -248,6 +247,21 @@ export const DeckGlWrapper = ({
   const selectedBaseMap = baseMaps.find((map) => map.id === selectedBaseMapId);
   const showTerrain = selectedBaseMap?.id === "Terrain";
   const mapStyle = selectedBaseMap?.mapUrl;
+
+  const [terrainTiles, setTerrainTiles] = useState({});
```
[the remark this answers is not in the document]
Comment (mspivak-actionengine) [reply src/components/deck-gl-wrapper/deck-gl-wrapper.tsx]: Done.

File: src/components/pick-pane-panel/pick-pane-panel.tsx [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+  cursor: pointer;
+  border-width: 0;
+  border-radius: 8px;
+
+  ${({ active = false }) =>
+    active &&
+    css`
+      box-shadow: 0 0 0 2px #000000, 0 0 0 4px #ffffff;
+      // -moz-box-shadow: 0 0 0 2px #000000, 0 0 0 4px #ffffff;
+      // -webkit-shadow: 0 0 0 2px #000000, 0 0 0 4px #ffffff;
```
[the remark this answers is not in the document]
Comment (mspivak-actionengine) [reply src/components/pick-pane-panel/pick-pane-panel.tsx]: Done.

File: src/components/deck-gl-wrapper/deck-gl-wrapper.tsx [hunk 9]
```diff
@@ -289,28 +303,26 @@ export const DeckGlWrapper = ({
...
-
-  const colorsByAttribute = useAppSelector(selectColorsByAttribute);
-
   const dispatch = useAppDispatch();
-
   /** Load debug texture if necessary */
   useEffect(() => {
-    if (loadDebugTextureImage && !uvDebugTexture) {
-      dispatch(fetchUVDebugTexture());
+//  useMemo(() => {
```
[the remark this answers is not in the document]
Comment (mspivak-actionengine) [reply src/components/deck-gl-wrapper/deck-gl-wrapper.tsx]: Done.

File: src/components/pick-pane-panel/pick-pane-panel.tsx (hunk 2 above)
Comment (mspivak-actionengine) [reply src/components/pick-pane-panel/pick-pane-panel.tsx]: Reworked completely

File: src/components/pick-pane-panel/pick-pane-panel.tsx (hunk 3 above)
Comment (mspivak-actionengine) [reply src/components/pick-pane-panel/pick-pane-panel.tsx]: Done.

File: src/components/debug-panel/debug-panel.tsx (hunk 1 above)
Comment (mspivak-actionengine) [reply src/components/debug-panel/debug-panel.tsx]: Removed the second

**Review by belom88 — CHANGES_REQUESTED** [review]
It is unnecessary complexity with textures initialization and adding textures via `uv-debug-texture-slice`. The responsibilities of slices are clear:
* `icon-list-slice` - manage lists;
* `uv-debug-texture-slice` - load the texture when the url has changed.

How they should work together:
* Init the list directly in `icon-list-slice`. It is OK that there will be "list of textures" in this slice initialized. It is not a generic component for a UI library. All add and remove list item logic should dispatch directly to this slice;
* `uv-debug-texture-slice` doesn't know about multiple textures. It stores only a loaded `ImageBitmap`. Add `url` argument to `fetchUVDebugTexture`. Make sure that the texture is null during requests. Probably, need to track request status (`idle`, `loading`, `succeded`, `failed`);
* We should have a `useEffect` in `DeckglWrapper` - it should dispatch `fetchUVDebugTexture` when selected texture has changed

File: src/components/bookmarks-panel/bookmarks-panel.tsx [hunk 10]
```diff
@@ -238,9 +241,10 @@ export const BookmarksPanel = ({
     setPopoverType(PopoverType.none);
   };
 
-  const onBookmarksUploadedHandler = (bookmarks) => {
+  const onBookmarksUploadedHandler = async (bookmarks) => {
```
[the remark this answers is not in the document]
Comment (mspivak-actionengine) [reply src/components/bookmarks-panel/bookmarks-panel.tsx]: Done.

File: src/components/icon-list-panel/icon-list-panel.tsx [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+  cursor: pointer;
+  border-width: 0;
+  border-radius: 8px;
+
+  ${({ active = false }) =>
+    active &&
+    css`
+      box-shadow: 0 0 0 2px #000000, 0 0 0 4px #ffffff;
+      -moz-box-shadow: 0 0 0 2px #000000, 0 0 0 4px #ffffff;
+      -webkit-shadow: 0 0 0 2px #000000, 0 0 0 4px #ffffff;
```
[the remark this answers is not in the document]
Comment (mspivak-actionengine) [reply src/components/icon-list-panel/icon-list-panel.tsx]: Done.

File: src/components/debug-panel/debug-panel.tsx [hunk 12]
```diff
@@ -159,6 +201,50 @@ export const DebugPanel = ({ onClose }: DebugPanelProps) => {
...
+                id: `${hash}`,
+                image: null,
+                imageUrl: url,
+                icon: url,
+                custom: true,
+              };
+              await dispatch(
+                addUVDebugTexture({ texture: texture, setCurrent: true })
+              );
+            }}
```
[the remark this answers is not in the document]
Comment (mspivak-actionengine) [reply src/components/debug-panel/debug-panel.tsx]: Done.

File: src/components/deck-gl-wrapper/deck-gl-wrapper.spec.tsx [hunk 13]
```diff
@@ -184,14 +183,10 @@ describe("Deck.gl I3S map component", () => {
     expect(controller2).toEqual(controllerExpected);
   });
 
-  it("Should load UV debug texture", () => {
+  it("Should load UV debug texture", async () => {
     const { rerender } = callRender(renderWithProvider, {
       loadDebugTextureImage: true,
     });
-    expect(load).toHaveBeenCalledWith(
```
[the remark this answers is not in the document]
Comment (mspivak-actionengine) [reply src/components/deck-gl-wrapper/deck-gl-wrapper.spec.tsx:191]: Done.

File: src/redux/store.ts [hunk 14]
```diff
@@ -38,7 +40,7 @@ export const setupStore = (preloadedState?: PreloadedState<RootState>) => {
           // Ignore these action types
           ignoredActions: ["fetchUVDebugTexture/fulfilled"],
           // Ignore these paths in the state
-          ignoredPaths: ["uvDebugTexture.value"],
+          ignoredPaths: ["uvDebugTexture.images"],
```
Comment (belom88) [inline src/redux/store.ts:43]: the path seems wrong

File: src/types.ts [hunk 15]
```diff
@@ -384,3 +384,27 @@ export enum FetchingStatus {
...
+
+export enum IconListSetName {
+  uvDebugTexture = "uvDebugTexture",
+  baseMap = "baseMap",
+}
+
+export interface IIconItem {
+  id: string;
+  icon: string;
+  group?: string;
```
Comment (belom88) [inline src/types.ts:396]: What the group?

File: src/utils/bookmarks-utils.ts [hunk 16]
```diff
@@ -162,3 +173,26 @@ export const checkBookmarksByPageId = (
 
   return pageId;
 };
+
+/**
+ * Parses json string containing bookmarks according to the schema.
+ * @param bookmarks
+ * @returns Array of bookmarks or null if the json string is invalid.
+ */
+export const parseBookmarks = async (
```
Comment (belom88) [inline src/utils/bookmarks-utils.ts:182]: Add a test

File: src/redux/slices/icon-list-slice.ts [hunk 17]
```diff
@@ -0,0 +1,178 @@
+import { createSlice, PayloadAction } from "@reduxjs/toolkit";
```
Comment (belom88) [inline src/redux/slices/icon-list-slice.ts:1]: Add test cases

File: src/components/deck-gl-wrapper/deck-gl-wrapper.tsx [hunk 18]
```diff
@@ -344,15 +349,15 @@ export const DeckGlWrapper = ({
     });
   }, [loadTiles]);
 
-  useEffect(() => {
+  useMemo(() => {
```
Comment (belom88) [inline src/components/deck-gl-wrapper/deck-gl-wrapper.tsx:352]: This change seems wrong. `useMemo` should return a value

**Review by belom88 — CHANGES_REQUESTED** [review]
image

* Upload panel cause horizontal scroll
* Custom texture doesn't work;
* Add gaps between title and drag&drop area and between "Cancel" button and drag&drop area;
* Add gaps between upload panel and parent debug panel

File: src/components/deck-gl-wrapper/deck-gl-wrapper.tsx [hunk 19]
```diff
@@ -310,18 +310,11 @@ export const DeckGlWrapper = ({
 
   const dispatch = useAppDispatch();
 
-  /** Load debug texture if necessary */
   useMemo(() => {
-    if (loadDebugTextureImage && !uvDebugTexture) {
```
[the remark this answers is not in the document]
Comment (mspivak-actionengine) [reply src/components/deck-gl-wrapper/deck-gl-wrapper.tsx:309]: Done.

File: src/components/debug-panel/debug-panel.tsx [hunk 20]
```diff
@@ -56,15 +68,52 @@ const RadioButtonWrapper = styled.div`
...
+const TextureControlPanel = styled.div`
+  display: flex;
+  flex-direction: column;
+  justify-content: start;
+  align-items: start;
+
+  margin-top: 0px;
+  margin-bottom: 0px;
+  margin-right: 16px;
+  margin-left: 16px;
```
[the remark this answers is not in the document]
Comment (mspivak-actionengine) [reply src/components/debug-panel/debug-panel.tsx]: reworked.

File: src/components/debug-panel/debug-panel.tsx [hunk 21]
```diff
@@ -159,6 +208,35 @@ export const DebugPanel = ({ onClose }: DebugPanelProps) => {
...
+            </ActionIconButton>
+          </TextureControlPanel>
+        )}
+
+        {showFileUploadPanel && (
+          <UploadPanel
+            title={"Upload Texture"}
+            dragAndDropText={"Drag and drop your texture file here"}
+            fileType={FileType.binary}
+            multipleFiles={true}
```
[the remark this answers is not in the document]
Comment (mspivak-actionengine) [reply src/components/debug-panel/debug-panel.tsx]: Done.

File: src/redux/slices/uv-debug-texture-slice.ts [hunk 22]
```diff
@@ -25,27 +25,35 @@ const uvDebugTextureSlice = createSlice({
...
+  const state = (getState() as RootState).uvDebugTexture;
+  const el = state.textureArray.find((item) => item.imageUrl === imageUrl);
 
-    return { value: image };
+  if (!el || !el.image) {
+    const image = (await load(imageUrl, ImageLoader)) as ImageBitmap;
+    return { imageUrl: imageUrl, image: image };
   }
-);
+  return { imageUrl: imageUrl, image: null };
```
[the remark this answers is not in the document]
Comment (mspivak-actionengine) [reply src/redux/slices/uv-debug-texture-slice.ts]: Done.

File: src/redux/slices/icon-list-slice.ts [hunk 23]
```diff
@@ -0,0 +1,178 @@
...
+}
+
+/**
+ * @example "BaseMap", "uvDebugTexture", "desktopBackground"
+ */
+interface IIconListState {
+  iconListSets: Record<string, IIconListSet>;
+}
+
+const initialState = (): IIconListState => {
```
[the remark this answers is not in the document]
Comment (mspivak-actionengine) [reply src/redux/slices/icon-list-slice.ts]: Done.

File: src/redux/slices/icon-list-slice.ts [hunk 24]
```diff
@@ -0,0 +1,178 @@
...
+      const arg = action.payload;
+      const sets = state.iconListSets[arg.iconListSetName];
+      const element = sets?.iconList.find((item) => item.id === arg.id);
+      // "custom" element only can be deleted.
+      if (element && element.custom) {
+        sets.iconList = sets.iconList.filter((item) => item.id !== arg.id);
+        if (sets.iconItemIdPicked === arg.id) {
+          sets.iconItemIdPicked = "";
+        }
+      }
```
[the remark this answers is not in the document]
Comment (mspivak-actionengine) [reply src/redux/slices/icon-list-slice.ts]: Done.

File: src/redux/store.ts (hunk 14 above)
Comment (mspivak-actionengine) [reply src/redux/store.ts]: Done.

File: src/types.ts (hunk 15 above)
Comment (mspivak-actionengine) [reply src/types.ts:401]: Added comments.

File: src/utils/bookmarks-utils.ts (hunk 16 above)
Comment (mspivak-actionengine) [reply src/utils/bookmarks-utils.ts:182]: Done.

File: src/redux/slices/icon-list-slice.ts (hunk 17 above)
Comment (mspivak-actionengine) [reply src/redux/slices/icon-list-slice.ts:1]: Done.

File: src/components/deck-gl-wrapper/deck-gl-wrapper.tsx (hunk 18 above)
Comment (mspivak-actionengine) [reply src/components/deck-gl-wrapper/deck-gl-wrapper.tsx]: Done.

File: src/components/icon-list-panel/icon-list-panel.tsx [hunk 25]
```diff
@@ -0,0 +1,91 @@
...
+                })
+              );
+            }}
+          />
+        ))}
+      </TexturePanel>
+    );
+  };
+
+  return renderTexturePanel(imageArray);
```
[the remark this answers is not in the document]
Comment (mspivak-actionengine) [reply src/components/icon-list-panel/icon-list-panel.tsx]: Done.

File: src/components/deck-gl-wrapper/deck-gl-wrapper.tsx [hunk 26]
```diff
@@ -295,12 +295,14 @@ export const DeckGlWrapper = ({
...
     selectIconItemPicked(IconListSetName.uvDebugTexture)
   );
-  const imageUrl = (iconItemPicked?.extData.imageUrl as string) || "";
+  const imageUrl = (iconItemPicked?.extData?.imageUrl as string) || "";
   const uvDebugTexture = useAppSelector(selectUVDebugTexture(imageUrl));
   const uvDebugTextureRef = useRef<ImageBitmap | null>(null);
   uvDebugTextureRef.current = uvDebugTexture;
   const [needTransitionToTileset, setNeedTransitionToTileset] = useState(false);
 
+  const [forceRefresh, setForceRefresh] = useState(false);
```
Comment (belom88) [inline src/components/deck-gl-wrapper/deck-gl-wrapper.tsx:304]: `forceRefresh` is changed but never used

File: src/redux/slices/uv-debug-texture-slice.ts [hunk 27]
```diff
@@ -25,16 +25,21 @@ const uvDebugTextureSlice = createSlice({
       fetchUVDebugTexture.fulfilled,
       (
         state: uvDebugTextureState,
-        action: PayloadAction<{ imageUrl: string; image: ImageBitmap | null }>
+        action: PayloadAction<{
+          imageUrl: string;
+          image: ImageBitmap | null;
```
Comment (belom88) [inline src/redux/slices/uv-debug-texture-slice.ts:30]: Does it make sense to push `null` to `image`?

**Review by belom88 — APPROVED** [review]

File: src/redux/slices/uv-debug-texture-slice.ts (hunk 27 above)
Comment (mspivak-actionengine) [reply src/redux/slices/uv-debug-texture-slice.ts:30]: Removed null from the type.

**Pull request merged**